python_modules/dagster/dagster/components/defs.py

```diff
@@ -0,0 +1,133 @@
+from functools import cached_property
+
+from dagster_shared import check
+
+from dagster._core.definitions.asset_check_spec import AssetCheckSpec
+from dagster._core.definitions.asset_checks import AssetChecksDefinition
+from dagster._core.definitions.asset_key import AssetCheckKey, AssetKey, CoercibleToAssetKey
+from dagster._core.definitions.asset_spec import AssetSpec
+from dagster._core.definitions.assets import AssetsDefinition
+from dagster._core.definitions.cacheable_assets import CacheableAssetsDefinition
+from dagster._core.definitions.definitions_class import Definitions
+from dagster._core.definitions.external_asset import create_external_asset_from_source_asset
+from dagster._core.definitions.source_asset import SourceAsset
+
+
+def canonicalize_assets_in_definitions(
+    definitions: Definitions,
+) -> tuple[
+    dict[AssetKey, AssetSpec],
+    dict[AssetKey, AssetsDefinition],
+    dict[AssetCheckKey, AssetChecksDefinition],
+]:
+    asset_specs: dict[AssetKey, AssetSpec] = {}
+    assets_defs: dict[AssetKey, AssetsDefinition] = {}
+    asset_checks_defs: dict[AssetCheckKey, AssetChecksDefinition] = {}
+
+    for asset_or_asset_check in list(definitions.assets or []) + list(
+        definitions.asset_checks or []
+    ):
+        if isinstance(asset_or_asset_check, CacheableAssetsDefinition):
+            raise ValueError(
+                "CacheableAssetsDefinitions are not allowed in components. Convert to AssetsDefinition"
+            )
+        elif isinstance(asset_or_asset_check, SourceAsset):
+            external_assets_def = create_external_asset_from_source_asset(asset_or_asset_check)
+            if asset_or_asset_check.is_observable:
+                assets_defs[asset_or_asset_check.key] = external_assets_def
+                asset_specs[asset_or_asset_check.key] = external_assets_def.get_asset_spec(
+                    asset_or_asset_check.key
+                )
+            else:
+                for key in external_assets_def.keys:
+                    asset_specs[key] = external_assets_def.get_asset_spec(key)
+        elif isinstance(asset_or_asset_check, AssetSpec):
+            asset_specs[asset_or_asset_check.key] = asset_or_asset_check
+        # AssetsDefinitions or AssetChecksDefinitions can be passed to either definitions.assets or definitions.asset_checks
+        elif isinstance(asset_or_asset_check, AssetChecksDefinition):
+            for key in asset_or_asset_check.check_keys:
+                asset_checks_defs[key] = asset_or_asset_check
+        elif isinstance(asset_or_asset_check, AssetsDefinition):
+            for key in asset_or_asset_check.keys:
+                assets_defs[key] = asset_or_asset_check
+                asset_specs[key] = asset_or_asset_check.get_asset_spec(key)
+        else:
+            check.failed(f"Unexpected asset type: {type(asset_or_asset_check)}")
+    return asset_specs, assets_defs, asset_checks_defs
+
+
+class Defs:
```

Comment:
A few naming spitballs since I find the relationship of `Defs` and `Definitions` pretty unclear:
- `ModuleDefs` - the counterpose of `DefsModule`, the defs contained in a module
- `ComponentDefs`
- `DefinitionsView` - since this is a read-oriented convenience layer on top of Definitions that you will never manipulate

Reply:
`ModuleDefs` looks good to me